pkg/processor/runtime/python/py/_nuclio_wrapper.py

```diff
@@ -231,39 +220,52 @@ async def _initialize_context(self):
                 raise
 
     def _register_to_signal(self):
-        on_termination_signal = functools.partial(self._on_termination_signal, Constants.termination_signal.name)
-        on_drain_signal = functools.partial(self._on_drain_signal, Constants.drain_signal.name)
-        on_continue_signal = functools.partial(self._on_continue_signal, Constants.continue_signal.name)
-
-        asyncio.get_running_loop().add_signal_handler(Constants.termination_signal, on_termination_signal)
-        asyncio.get_running_loop().add_signal_handler(Constants.drain_signal, on_drain_signal)
-        asyncio.get_running_loop().add_signal_handler(Constants.continue_signal, on_continue_signal)
+        signal.signal(signal.SIGUSR1, self._on_termination_signal)
+        signal.signal(signal.SIGUSR2, self._on_drain_signal)
+        signal.signal(signal.SIGCONT, self._on_continue_signal)
 
-    def _on_drain_signal(self, signal_name):
+    def _on_drain_signal(self, signal_number, frame):
         # do not perform draining if discarding events
         if self._discard_events:
             self._logger.debug('Draining signal is received, but it will be ignored as the worker is already drained')
             return
 
-        self._logger.debug_with('Received signal', signal=signal_name)
-        self._is_drain_needed = True
+        self._logger.debug_with('Received signal, calling draining callback',
+                                signal=signal.Signals(signal_number).name)
+
+        if self._is_waiting_for_event:
+            self._logger.debug('Wrapper is waiting for an event, calling drain handler')
 
+            # call the drain handler here as the event loop is stuck waiting for an event
+            self._call_drain_handler()
+        else:
+            self._logger.debug('Wrapper is handling an event, setting drain flag to true')
+
+            # set the flag to true so the event loop will call the drain handler
+            # after the current event is handled
+            self._is_drain_needed = True
         # set the flag to True to stop processing events which are received after draining
         self._discard_events = True
```

Comment:
I would move this before the if-else, so events will be discarded while the drain handler is invoked.